### PR TITLE
CORTX-32039: Add rgw core file clean up script as part of system's logrotate directory.

### PR DESCRIPTION
**Pull request by sachitanands**
Signed-off-by: Sachitanand Shelake <sachitanand.shelake@seagate.com>

# Problem Statement
- Handle clean up of rgw core files using /etc/logrotate.d scripts.

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Review by priyanka25081999 — APPROVED** [review]
Looks good!

File: src/rgw/setup/rgw.py [hunk 1]
```diff
@@ -827,6 +827,22 @@ def _logrotate_generic(conf: MappedConf):
             Log.info(f'{const.LOGROTATE_TMPL} file copied to {const.LOGROTATE_CONF}')
         except Exception as e:
             Log.error(f"Failed to configure logrotate for {const.COMPONENT_NAME}. ERROR:{e}")
+
+        # Copy log rotate script for rgw core files into system's logrotate directory.
+        old_core_logrotate_file = os.path.join(const.LOGROTATE_DIR, const.CORE_LOGROTATE_FILE_NAME)
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:832]: why handling these old files? any fresh deployment wont have old files right.

File: src/rgw/setup/rgw.py [hunk 2]
```diff
@@ -827,6 +827,22 @@ def _logrotate_generic(conf: MappedConf):
...
+        core_dir_path = os.path.join(log_file_path, const.RGW_CORE_FILE_DIR_NAME)
+        if os.path.exists(old_core_logrotate_file):
+            os.remove(old_core_logrotate_file)
+        try:
+            with open(const.CORE_LOGROTATE_TMPL, 'r') as f:
+                content = f.read()
+            content = content.replace('TEMP_CORE_FILE_DIR_PATH', core_dir_path)
+            with open(const.CORE_LOGROTATE_FILE, 'w') as f:
+                f.write(content)
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:841]: why this TEMP_CORE_FILE_DIR_PATH maintained in file, which needs to be replaced then

File: src/rgw/setup/templates/rgw_core_logrotate.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            # get the files sorted by time modified (most recently modified comes last),
+            # that is oldest files will come on top.
+            files_to_remove=`ls -tr "$rgw_core_dir" | grep core | head -n $remove_file_count`
+            for file in $files_to_remove
+            do
+              rm -f "$rgw_core_dir/$file"
+            done
+            echo "## deleted ($remove_file_count) core file(s) from directory: $rgw_core_dir ##"
+        else
+            echo "## No rgw core files to remove ##"
```
Comment (cdeshmukh) [inline src/rgw/setup/templates/rgw_core_logrotate.sh:81]: echo on line 81/84/87 can fill up the console when no core files seen. keep it silent, we dont need prints in such cases

**Review by cdeshmukh — COMMENTED** [review]
Also address codacy issues

File: src/rgw/setup/rgw.py (hunk 1 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py:832]: added for additional validation. but yes your point is valid, we can safely remove this.

File: src/rgw/setup/rgw.py (hunk 2 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py:841]: this is place holder and kept in this template file. 
During mini-provisioner, the actual path of rgw core file will be replaced here.
This is simillar logic we followed for rgw log rotation code.
this way we are ignoring any hardcoding of log path in our code.

File: src/rgw/setup/rgw.py (hunk 1 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py]: done

File: src/rgw/setup/templates/rgw_core_logrotate.sh (hunk 3 above)
Comment (sachitanands) [reply src/rgw/setup/templates/rgw_core_logrotate.sh]: done. removed this else block itself

File: src/rgw/setup/rgw.py (hunk 2 above)
Comment (cdeshmukh) [reply src/rgw/setup/rgw.py:838]: ok

File: src/rgw/setup/templates/rgw_core_logrotate.sh [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+usage() { echo "Usage: bash $(basename "$0")[--help|-h]
+                   [-n maxRgwCoreFileCount]
+Retain recent generated rgw core files of given count and remove rest of core files.
+
+where:
+-n            number of latest rgw core files to retain (Default count for core files is 5)
+--help|-h     display this help and exit" 1>&2; exit 1; }
+
+# max rgw core files count in each rgw log directory
+core_files_max_count=5
```
Comment (cdeshmukh) [inline src/rgw/setup/templates/rgw_core_logrotate.sh:36]: 5 cores by default would be high, given one core file size can have atleast 500MB size, lets keep 2

Comment (sachitanands) [reply src/rgw/setup/templates/rgw_core_logrotate.sh:36]: okay.will update it

Comment (sachitanands) [reply src/rgw/setup/templates/rgw_core_logrotate.sh]: done. added default value as 2

File: src/rgw/setup/templates/rgw_core_logrotate.sh [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            done
+            echo "## deleted ($remove_file_count) core file(s) from directory: $rgw_core_dir ##"
+        fi
+     fi
+else
+    echo "ERROR !!! Invalid directory path : $rgw_core_dir. Please add proper path of rgw core directory."
+    exit
+fi
+
+echo "Done with rgw core log rotation script."
```
Comment (cdeshmukh) [inline src/rgw/setup/templates/rgw_core_logrotate.sh:87]: "Done with rgw core file rotation"

Comment (sachitanands) [reply src/rgw/setup/templates/rgw_core_logrotate.sh]: done

**Review by shriya-deshmukh — APPROVED** [review]

**Review by cdeshmukh — APPROVED** [review]

**Pull request merged**